# IMPORTS

In [1]:
import sys
sys.path.append('../src/')
import os
import copy

import numpy as np
import tensorly as tn

from loewnerize_acts import loewnerize_acts


## get matlab engine set up with tensorlab


In [2]:
import matlab.engine
eng = matlab.engine.start_matlab()
s = eng.genpath('~/ware/matlab_tools/tensorlab')
eng.addpath(s, nargout=0)

# mess around
## load activations for 300 node layer (inputs - x & outputs - f(x))

In [2]:
targets_path = '../data/Lenet300100/activations/test/targets.npy'
targets = np.load(targets_path)

inputs_path = '../data/Lenet300100/activations/test/99/Linear_0.npy'
inputs_raw = np.load(inputs_path)

outputs_path = '../data/Lenet300100/activations/test/99/Rational_1.npy'
outputs_raw = np.load(outputs_path)

tensorlab_path = '~/ware/matlab_tools/tensorlab'

In [3]:
# loewnerize activations --> returns a list of activation tensors split according to targets (0-9)
lwn_tensors = loewnerize_acts(inputs_raw, outputs_raw, targets, tensorlab_path)

ValueError: setting an array element with a sequence.

In [17]:
print(lwn_tensors[0].shape)
print(type(lwn_tensors[0]))
print(type(lwn_tensors))

(150, 150, 979)
<class 'numpy.ndarray'>
<class 'list'>


In [15]:
datapath = '../data/Lenet300100/activations/test/99/ten_acts/300_node_act_tens'
np.save(datapath, np.asarray(lwn_tensors))

ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 3 dimensions. The detected shape was (10, 150, 150) + inhomogeneous part.

In [5]:
targets.shape


(9984,)

In [6]:
inputs_raw.shape

(39, 256, 300)

In [7]:
outputs_raw.shape

(39, 256, 300)

### 39 batchs of 256 examples = 9984 total examples
Need to reshape inputs and outputs 

In [8]:
inputs = inputs_raw.reshape(-1, inputs_raw.shape[-1])
outputs = outputs_raw.reshape(-1, outputs_raw.shape[-1])

In [9]:
inputs.shape[1]

300

In [8]:
targets.ndim

1

In [15]:
i_0 = np.empty((0,300))
i_1 = np.empty((0,300))
i_2 = np.empty((0,300))
i_3 = np.empty((0,300))
i_4 = np.empty((0,300))
i_5 = np.empty((0,300))
i_6 = np.empty((0,300))
i_7 = np.empty((0,300))
i_8 = np.empty((0,300))
i_9 = np.empty((0,300))

o_0 = np.empty((0,300))
o_1 = np.empty((0,300))
o_2 = np.empty((0,300))
o_3 = np.empty((0,300))
o_4 = np.empty((0,300))
o_5 = np.empty((0,300))
o_6 = np.empty((0,300))
o_7 = np.empty((0,300))
o_8 = np.empty((0,300))
o_9 = np.empty((0,300))

for idx,x in enumerate(targets):
    if x == 0:
        i_0 = np.vstack([i_0,inputs[idx,:]])
        o_0 = np.vstack([o_0,outputs[idx,:]])
    elif x == 1:
        i_1 = np.vstack([i_1,inputs[idx,:]])
        o_1 = np.vstack([o_1,outputs[idx,:]])
    elif x == 2:
        i_2 = np.vstack([i_2,inputs[idx,:]])
        o_2 = np.vstack([o_2,outputs[idx,:]])
    elif x == 3:
        i_3 = np.vstack([i_3,inputs[idx,:]])
        o_3 = np.vstack([o_3,outputs[idx,:]])
    elif x == 4:
        i_4 = np.vstack([i_4,inputs[idx,:]])
        o_4 = np.vstack([o_4,outputs[idx,:]])
    elif x == 5:
        i_5 = np.vstack([i_5,inputs[idx,:]])
        o_5 = np.vstack([o_5,outputs[idx,:]])
    elif x == 6:
        i_6 = np.vstack([i_6,inputs[idx,:]])
        o_6 = np.vstack([o_6,outputs[idx,:]])
    elif x == 7:
        i_7 = np.vstack([i_7,inputs[idx,:]])
        o_7 = np.vstack([o_7,outputs[idx,:]])
    elif x == 8:
        i_8 = np.vstack([i_8,inputs[idx,:]])
        o_8 = np.vstack([o_8,outputs[idx,:]])
    elif x == 9:
        i_9 = np.vstack([i_9,inputs[idx,:]])
        o_9 = np.vstack([o_9,outputs[idx,:]])

In [16]:
print(o_0.shape)
print(o_1.shape)
print(o_2.shape)
print(o_3.shape)
print(o_4.shape)
print(o_5.shape)
print(o_6.shape)
print(o_7.shape)
print(o_8.shape)
print(o_9.shape)

(979, 300)
(1133, 300)
(1030, 300)
(1008, 300)
(980, 300)
(890, 300)
(956, 300)
(1027, 300)
(973, 300)
(1008, 300)


### Lownerize '0' examples

In [19]:
num_rows = o_0.shape[0]
lwn_0 = np.empty((150,150,0))
for i in range(0,num_rows):
    lwn_tmp = eng.loewnerize(o_0[i,:], 'T', i_0[i,:])
    lwn_0 = np.dstack((lwn_0,lwn_tmp))
print(lwn_0.shape)

# now apply block-term decomp into (3,3,1) terms
L = 3 * np.ones(10, dtype=int)
print(L)


(150, 150, 979)
[3 3 3 3 3 3 3 3 3 3]


In [20]:
Uhat = eng.ll1(lwn_0, L)

In [21]:
res = eng.ll1res(lwn_0, Uhat)
print(res)

> In fmt (line 211)
In btdres (line 38)
In ll1res (line 56)


TypeError: unsupported data type returned from MATLAB